src/Rules/Classes/EnumSanityRule.php

```diff
@@ -122,6 +124,81 @@ public function processNode(Node $node, Scope $scope): array
 			}
 		}
 
+		$enumCases = [];
+		foreach ($node->stmts as $stmt) {
+			if (!$stmt instanceof Node\Stmt\EnumCase) {
+				continue;
+			}
+			$caseName = $stmt->name->name;
+
+			if (($stmt->expr instanceof Node\Scalar\LNumber || $stmt->expr instanceof Node\Scalar\String_)) {
+				if ($node->scalarType === null) {
+					$errors[] = RuleErrorBuilder::message(sprintf(
+						'Enum %s is not backed, but %s::%s has value %s',
+						$node->namespacedName->toString(),
+						$node->namespacedName->toString(),
+						$caseName,
+						$stmt->expr->value,
+					))
+						->identifier('enum.nonBackedEnumHasValue')
+						->line($stmt->getLine())
+						->nonIgnorable()
+						->build();
+				} else {
+					$caseValue = $stmt->expr->value;
+
+					if (!isset($enumCases[$caseValue])) {
+						$enumCases[$caseValue] = [];
+					}
+
+					$enumCases[$caseValue][] = $caseName;
+				}
+			}
+
+			if ($node->scalarType?->name === 'int' && !($stmt->expr instanceof Node\Scalar\LNumber)) {
+				$errors[] = RuleErrorBuilder::message(sprintf(
+					"Enum case %s::%s doesn't match the 'int' type",
+					$node->namespacedName->toString(),
+					$caseName,
+				))
+					->identifier('enum.caseTypeMismatch')
+					->line($stmt->getLine())
+					->nonIgnorable()
+					->build();
+			}
+
+			$isStringBackedWithoutStringCase = $node->scalarType?->name === 'string' && !($stmt->expr instanceof Node\Scalar\String_);
+			if (!$isStringBackedWithoutStringCase) {
+				continue;
+			}
+			$errors[] = RuleErrorBuilder::message(sprintf(
+				"Enum case %s::%s doesn't match the 'string' type",
+				$node->namespacedName->toString(),
+				$caseName,
+			))
+				->identifier('enum.caseTypeMismatch')
+				->line($stmt->getLine())
+				->nonIgnorable()
+				->build();
+		}
```

Comment:
Here I had to use early return (`if (!$isStringBackedWithoutStringCase) {`) because linter forced me to do that.

Reply:
You might want to look into fine-tuning https://github.com/slevomat/coding-standard/blob/master/doc/control-structures.md#slevomatcodingstandardcontrolstructuresearlyexit-

I obviously can't speak for Ondřej, but the last time a similar thing like this was discussed it sounded like he is open to look into adapting it.